content/blog/2019-10-23-loop-unrolling-with-bril.md

```diff
@@ -0,0 +1,154 @@
+
+### Experiment Results
+
+We use the hand-written test programs to verify the correctness of our loop unrolling passes and measure performance improvements. The functioanl correctness of the unrolled code can be verified with bril intepreter. To evaluate the performance between original and unrolled programs, we converted the bril code from json representation to C using the Bril C back end. We use GCC to compile the C program without any optimization (``-O0``), run it for 10000 times with bash loop to measure the total time consumption. Experiments are performed on a server with an 2.20GHz Intel Xeon processor and 128GB memory. All programs are single-thread. The benchmarks can be found [here](https://github.com/seanlatias/bril/tree/arrays/test/unroll).
+
+#### Static Analysis Evaluation
+The program is located [here](https://github.com/seanlatias/bril/blob/arrays/examples/loop_unroll.py). To execute the pass, run the following command.
+
+``python loop_unroll.py output.json < input.json``
+
+The pass can successfully unroll the loops in our test programs. For the "double loop" test case, the pass correctly identifies the inner loop and unrolls it. Execution time of the original and unrolled versions are as follows:
+
+| Benchmark | Original (s) | Unrolled (s) |
+|:---------:|:------------:|:------------:|
+|Two consecutive loops|3.530|3.621|
+|Nested loop|3.509|3.533|
+|Vector sum|3.599|3.508|
+|Set vector value|3.473|3.671|
+
+There is no significant performance difference between the unrolled version and the original version. One reason might be 
+because the loop tripcounts in the test programs are all very small. As a result, we modify the vector sum program to have 
+large tripcount (2000) and compare the performance again. To enable this we modify our loop unrolling pass to allow unrolling very large loops. In terms of the sizes of the executables, the unrolled version is 12.7x larger than the original version. The original and unrolled program uses 3.693s and 3.755s, repectively. There is still no significant performance benefit. The possible reason is that the control flow of our test examples is too regular and can be perfectly predicted by the branch predictor in modern processors. Since the processor is pipelined, we do not observe significant performance improvement even after we remove the redundant branches after unrolling. For most benchmarks we actually observe slowdowns, which possibly result from instruction cache misses caused by larger executable sizes. 
```

Comment:
Another option is that, even at `-O0` optimization, GCC is causing your unrolled programs to do somewhat unpredictable things.